File: www/core/imager/factory.ts
import { HotelsCode, hotelsConfig, type THotelsCode } from "@/config/hotels";
import { Imager } from "./imager";
import { HabbletImagerProvider, HabbliveImagerProvider, type ImagerProvider } from "./providers";

export abstract class ImagerFactory {
  public static create(hotel: THotelsCode): Imager {
    let provider: ImagerProvider;

    switch (hotel) {
      case HotelsCode.Habblet:
        provider = new HabbletImagerProvider(
          hotelsConfig[hotel].imagerUrl,
          hotelsConfig[hotel].apiBaseUrl,
        );
        break;

      case HotelsCode.Habblive:
        provider = new HabbliveImagerProvider(hotelsConfig[hotel].imagerUrl);
        break;
    }

    return new Imager(provider);
  }
}

File: src/infra/imager/factory.rs
use crate::configs::hotels::{AvailableHotel, HOTELS_CONFIG};
use crate::error::SamambaiaError;
use crate::infra::imager::Imager;
use crate::infra::imager::providers::{
    HabbletImagerProvider,
    HabbliveImagerProvider,
    ImagerProvider,
};

pub struct ImagerFactory;

impl ImagerFactory {
    pub fn get_imager() -> Result<Imager, SamambaiaError> {
        let hotels_config = HOTELS_CONFIG.get_active_hotel_info();

        let provider = match HOTELS_CONFIG.active_hotel {
            AvailableHotel::Habblet => {
                HabbletImagerProvider::new(hotels_config.api_base_url, hotels_config.imager_url)
                    .map(|provider| Box::new(provider) as Box<dyn ImagerProvider>)
            }
            AvailableHotel::Habblive => Ok(Box::new(HabbliveImagerProvider::new(
                hotels_config.imager_url,
            )) as Box<dyn ImagerProvider>),
        }?;

        Ok(Imager::new(provider))
    }
}
